src/spawner.rs

```diff
@@ -1,77 +1,83 @@
+//! There you can find traits that are necessary for implementing executor.
+//! They are mostly unsafe, as we can't ensure the executor allows intended manipulations
+//! without causing UB.
 use futures::Future;
 
-pub trait Spawner<T> {
+pub unsafe trait Spawner<T> {
     type FutureOutput;
-    type SpawnHandle: Future<Output=Self::FutureOutput> + Send;
-    fn spawn<F: Future<Output=T> + Send + 'static>(f: F) -> Self::SpawnHandle;
+    type SpawnHandle: Future<Output = Self::FutureOutput> + Send;
+    fn spawn<F: Future<Output = T> + Send + 'static>(&self, f: F) -> Self::SpawnHandle;
 }
 
-pub trait FuncSpawner<T> {
+pub unsafe trait FuncSpawner<T> {
     type FutureOutput;
     type SpawnHandle: Future<Output = Self::FutureOutput> + Send;
-    fn spawn_func<F: FnOnce() -> T + Send + 'static>(f: F) -> Self::SpawnHandle;
+    fn spawn_func<F: FnOnce() -> T + Send + 'static>(&self, f: F) -> Self::SpawnHandle;
 }
 
-pub trait Blocker {
-    fn block_on<T, F: Future<Output=T>>(f: F) -> T;
+pub unsafe trait Blocker {
+    fn block_on<T, F: Future<Output = T>>(&self, f: F) -> T;
 }
 
 #[cfg(feature = "use-async-std")]
 pub mod use_async_std {
     use super::*;
-    use async_std::task::{spawn, block_on, spawn_blocking, JoinHandle};
+    use async_std::task::{block_on, spawn, spawn_blocking, JoinHandle};
+
+    #[derive(Default)]
     pub struct AsyncStd;
 
-    impl<T: Send + 'static> Spawner<T> for AsyncStd {
+    unsafe impl<T: Send + 'static> Spawner<T> for AsyncStd {
         type FutureOutput = T;
         type SpawnHandle = JoinHandle<T>;
 
-        fn spawn<F: Future<Output=T> + Send + 'static>(f: F) -> Self::SpawnHandle {
+        fn spawn<F: Future<Output = T> + Send + 'static>(&self, f: F) -> Self::SpawnHandle {
```

Comment:
Can you explain the reason behind adding `&self` and making it stateful?   It does feel more flexible, than just using the type, but was wondering if you had a concrete reason for this...

Reply:
There are cases, where you have multiple runtimes in the single process, with the variables pointing to them. Without making it stateful, such usecases are harder to implement, if even possible..

It is taken directly from our production use-case, so it's linked with the real world :)